File: results/0.25/Modbus/modbus-gpt-4-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// Function Prototypes
static void parse_modbus(const char *filename);
static void print_hex(const uint8_t *data, size_t length);

// Parsers
static HParser *modbus_tcp;
static HParser *modbus_pdu;
static HParser *modbus_function;
static HParser *modbus_data;

void build_modbus_parsers() {
    HParser *uint8 = h_uint8();
    HParser *uint16 = h_uint16();

    // MBAP Header
    HParser *transaction_id = uint16;
    HParser *protocol_id = uint16;
    HParser *length = uint16;
    HParser *unit_id = uint8;

    // Function Code and Data
    modbus_function = uint8;
    modbus_data = h_length_value(uint16, h_bytes(1));  // Use h_bytes for variable length data

    // Modbus PDU
    modbus_pdu = h_sequence(modbus_function, modbus_data, NULL);

    // Modbus TCP Frame (MBAP + PDU)
    modbus_tcp = h_sequence(transaction_id, protocol_id, length, unit_id, modbus_pdu, NULL);
}

void parse_modbus(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        fprintf(stderr, "Failed to open file: %s\n", filename);
        return;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t *buffer = malloc(file_size);
    if (!buffer) {
        fprintf(stderr, "Failed to allocate memory for file buffer\n");
        fclose(file);
        return;
    }

    fread(buffer, 1, file_size, file);
    fclose(file);

    HParseResult *result = h_parse(modbus_tcp, buffer, file_size);
    if (result) {
        printf("Parse successful!\n");
        print_hex(buffer, file_size);
    } else {
        printf("Parse failed!\n");
    }

    free(buffer);
}

void print_hex(const uint8_t *data, size_t length) {
    for (size_t i = 0; i < length; i++) {
        printf("%02X ", data[i]);
    }
    printf("\n");
}

int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <modbus_binary_file>\n", argv[0]);
        return 1;
    }

    build_modbus_parsers();
    parse_modbus(argv[1]);

    return 0;
}

File: results/0.25/Modbus/modbus-gpt-4-turbo.rs
use nom::{
    bytes::complete::{take, take_while_m_n},
    number::complete::{be_u16, be_u8},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::{self, Read};

#[derive(Debug)]
struct ModbusTCPFrame {
    transaction_id: u16,
    protocol_id: u16,
    length: u16,
    unit_id: u8,
    function_code: u8,
    data: Vec<u8>,
}

fn parse_modbus_tcp_frame(input: &[u8]) -> IResult<&[u8], ModbusTCPFrame> {
    let (input, transaction_id) = be_u16(input)?;
    let (input, protocol_id) = be_u16(input)?;
    let (input, length) = be_u16(input)?;
    let (input, unit_id) = be_u8(input)?;
    let (input, function_code) = be_u8(input)?;
    let (input, data) = take(length as usize - 2)(input)?; // Subtract 2 for unit_id and function_code

    Ok((
        input,
        ModbusTCPFrame {
            transaction_id,
            protocol_id,
            length,
            unit_id,
            function_code,
            data: data.to_vec(),
        },
    ))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: {} <file_path>", args[0]);
        return Ok(());
    }

    let file_path = &args[1];
    let mut file = File::open(file_path)?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_modbus_tcp_frame(&buffer) {
        Ok((_remaining, frame)) => {
            println!("{:?}", frame);
        }
        Err(error) => {
            println!("Failed to parse Modbus TCP frame: {:?}", error);
        }
    }

    Ok(())
}